contracts/EpochTokenLocker.sol

```diff
@@ -0,0 +1,146 @@
+pragma solidity ^0.5.0;
+
+import "openzeppelin-solidity/contracts/token/ERC20/ERC20.sol";
+import "openzeppelin-solidity/contracts/math/Math.sol";
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+
+
+// EpochTokenLocker saveguards tokens for applications with constant-balances during discrete epochs
+// It allows to deposit token which become credited in the next epoch and allows to request a token-withdraw
+// which becomes claimable after the current epoch expired.
+
+contract EpochTokenLocker {
+    using SafeMath for uint;
+
+    event Deposit(
+        address user,
+        address token,
+        uint amount,
+        uint stateIndex
+    );
+
+    event withdrawRequest(
+        address user,
+        address token,
+        uint amount,
+        uint stateIndex
+    );
+
+    event Withdraw(
+        address user,
+        address token,
+        uint amount
+    );
+
+    // User => Token => BalanceState
+    mapping(address => mapping(address => BalanceState)) private balanceStates;
+
+    struct BalanceState {
+        uint256 balance;
+        PendingFlux pendingDeposits; // deposits will be credited in any next epoch, i.e. currentStateIndex > stateIndex
+        PendingFlux pendingWithdraws; // withdraws are allowed in any next epoch , i.e. currentStateIndex > stateIndex
+    }
+
+    struct PendingFlux {
+        uint256 amount;
+        uint256 stateIndex;
+    }
+
+    uint256 public currentStateIndex = 0;
+
+    function updateAndGetBalance(address user, address token) public returns(uint256) {
+        updateDepositsBalance(user, token);
+        uint balance = balanceStates[user][token].balance;
```

Comment:
To my understanding, the optimizer is already smart enough to deal with these kinds of things.
I am wondering, probably loading balaceStates[user][token] would also require loading 5 variables, instead of just loading the 3 ones we really need. (If the optimizer is not smart enough)...

Reply:
I didn't understand last part.
Yep, it would make sense that the optimizer do that. In any case, I think it would be also more readable imo.

Reply:
balaceStates[user][token] is a struct with 5 variables. If I would define it inside the smart contract, it might have the impact that all 5 variables need to be read (5*500 gas) instead of just the 3 that are needed. 
But I am not sure, how good the optimizer is. Maybe he changes the code so that only 3 are read.

Reply:
Not sure, I would think that saving it in a intermediary variable would only save the reference to the start of the struct, so my intuition would say no variable would be loaded (we just save the pointer)

The we do the same reads using the pointer as the start position, and the fields would tell where to start reading.

It's usually a good practice to do this small refactoring for every programming language (specially for readability, but also a bit more performant), I would be surprised if in solidity is bad. Not sure how to make sure, it's not super important, but I'm sure this is something that is repeated in all the contracts since we will be working a lot with arrays of structs.